File: viewer/lib/core/leveling.dart
/// Cálculo de níveis alinhado a [app/core/leveling.py] (reward_level).
///
/// Nível 1: 50 pontos para avançar; depois threshold(N) = ceil(threshold(N-1) * 1.2)
/// via inteiro (x*6+4)//5.
library;

const int kBaseLevelThreshold = 50;

const int _growthNum = 6;
const int _growthDen = 5;

/// Progresso no nível atual para exibir na barra XP.
class LevelProgress {
  const LevelProgress({
    required this.level,
    required this.levelPoints,
    required this.nextThreshold,
  });

  final int level;
  final int levelPoints;
  final int nextThreshold;
}

int thresholdForLevel(int level) {
  if (level < 1) {
    throw ArgumentError.value(level, 'level', 'deve ser >= 1');
  }
  var threshold = kBaseLevelThreshold;
  for (var l = 2; l <= level; l++) {
    threshold = (threshold * _growthNum + (_growthDen - 1)) ~/ _growthDen;
  }
  return threshold;
}

LevelProgress computeLevelFromTotalPoints(int totalPoints) {
  var total = totalPoints < 0 ? 0 : totalPoints;
  var level = 1;
  var remaining = total;

  while (true) {
    final nextThreshold = thresholdForLevel(level);
    if (remaining >= nextThreshold) {
      remaining -= nextThreshold;
      level += 1;
      continue;
    }
    return LevelProgress(
      level: level,
      levelPoints: remaining,
      nextThreshold: nextThreshold,
    );
  }
}

/// Lê int a partir de JSON (int ou double).
int? readIntFromJson(dynamic value) {
  if (value == null) return null;
  if (value is int) return value;
  if (value is num) return value.toInt();
  return null;
}

/// Usa o total `points` da API como fonte da verdade (mesmo que level_* venham ausentes ou defasados).
LevelProgress levelProgressFromUserPointsMap(Map<String, dynamic> map) {
  final total = readIntFromJson(map['points']) ?? 0;
  return computeLevelFromTotalPoints(total);
}

File: app/core/leveling.py
"""Cálculo de níveis ("reward_level") baseado no total de pontos do usuário.

Regra:
- Nível 1: threshold para avançar = 50 pontos.
- A cada nível: threshold(N) = ceil(threshold(N-1) * 1.2).
  (Implementado com aritmética inteira: ceil(x * 6/5) = (x*6 + 4)//5)
- Ao atingir threshold do nível atual, usuário sobe e o contador "level_points"
  recebe carry over (sobra continua valendo no próximo nível).
"""

from __future__ import annotations

from typing import Tuple


BASE_LEVEL_THRESHOLD: int = 50
GROWTH_NUM: int = 6
GROWTH_DEN: int = 5


def threshold_for_level(level: int) -> int:
    """Retorna quantos pontos são necessários para sair do `level` e ir para o `level+1`."""
    if level < 1:
        raise ValueError("level deve ser >= 1")

    threshold = BASE_LEVEL_THRESHOLD
    # level=1 -> BASE; level=2 -> ceil(BASE*1.2) etc.
    for _ in range(2, level + 1):
        threshold = (threshold * GROWTH_NUM + (GROWTH_DEN - 1)) // GROWTH_DEN
    return threshold


def compute_level_from_total_points(total_points: int) -> Tuple[int, int, int]:
    """Calcula (level, level_points, next_threshold) a partir do total acumulado.

    `level_points` é a pontuação acumulada dentro do nível atual (carry over),
    que deve ficar sempre em [0, next_threshold - 1] (exceto quando threshold muda).
    """
    if total_points < 0:
        total_points = 0

    level = 1
    remaining = total_points

    while True:
        next_threshold = threshold_for_level(level)
        if remaining >= next_threshold:
            remaining -= next_threshold
            level += 1
            continue
        return level, remaining, next_threshold

